### PR TITLE
Support QoS in api_server

### PR DESCRIPTION
**Pull request by sallyjunjun**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily receiving feedbacks. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

The purpose of this commit is to introduce Quality of Service (QoS) functionality. When the system is in an overloaded state, this feature ensures prioritization based on user priority, allowing high-priority users to have precedence. For users with the same priority, resources can be utilized in accordance with the specified ratio.

## Modification

1. Added the QoS (Quality of Service) module along with configuration file parsing.
2. Introduced three rate-limiting interfaces: chat_completions_v1_qos, completions_v1_qos, and chat_interactive_v1_qos.
3. Included test files for this module.
4. No impact on the functionality of existing interfaces.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit tests to ensure the correctness.
3. If the modification has a dependency on downstream projects of a newer version, this PR should be tested with all supported versions of downstream projects.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

**Comment by lvhan028** [comment]
May provide user guide

**Comment by lvhan028** [comment]
Please resolve the linting error

File: lmdeploy/serve/qos_engine/qos_engine.py [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+        request_event = (request,event)
+        self.qos_user_group.enqueue(request_event)
+
+        await event.wait()
+
+        result_generator = await self.generate(request)
+        
+        # release self.availSlots resources
+        with self.lock:
+            if hasattr(request,'prompt'):
```
Comment (AllentDan) [inline lmdeploy/serve/qos_engine/qos_engine.py:135]: 要不改成类型判断吧，interactive 的请求，prompt 不是 List[str] 形式，而是 List[Dict[str, str]]。后者是一次对话的历史记录。目前只有 /v1/completions 接口会接受多个问题。

File: lmdeploy/serve/qos_engine/qos_engine.py [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+        except:
+            self.is_qos_enabled = False
+            self.user_id_maps = dict()
+            self.user_group_prio = []
+        logger.debug(f"is_qos_enabled: {self.is_qos_enabled}")
+        logger.debug(f"user_id_maps:  {self.user_id_maps}")
+        logger.debug(f"user_group_prio: {self.user_group_prio}")
+
+class QosEngine:
+    def __init__(self, instance_num=16, qos_tag="", engine=None, **kwargs) -> None:
```
Comment (AllentDan) [inline lmdeploy/serve/qos_engine/qos_engine.py:32]: 应该不用传 instance_num了，可以从engine拿。engine 默认参数应该也用不了

File: lmdeploy/serve/openai/api_server.py [hunk 3]
```diff
@@ -230,6 +365,167 @@ async def completion_stream_generator() -> AsyncGenerator[str, None]:
...
+    #         sequence_end=True,
+    #         request_output_len=request.max_tokens
+    #         if request.max_tokens else 512,
+    #         stop=False,
+    #         top_p=request.top_p,
+    #         temperature=request.temperature,
+    #         repetition_penalty=request.repetition_penalty,
+    #         ignore_eos=request.ignore_eos,
+    #         do_preprocess=False)
+    #     generators.append(result_generator)
```
Comment (AllentDan) [inline lmdeploy/serve/openai/api_server.py:432]: 可以删掉？

Comment (awslshadowstar) [reply lmdeploy/serve/openai/api_server.py:432]: 好的好的

File: lmdeploy/serve/openai/api_server.py [hunk 4]
```diff
@@ -551,6 +912,7 @@ def serve(model_path: str,
         allow_methods (List[str]): a list of allowed HTTP methods for CORS
         allow_headers (List[str]): a list of allowed HTTP headers for CORS
         log_level(str): set log level whose value among [CRITICAL, ERROR, WARNING, INFO, DEBUG]
+        qos_config_path (str): qos policy config path
```
[the remark this answers is not in the document]
Comment (awslshadowstar) [reply lmdeploy/serve/openai/api_server.py:915]: 如果没有的话，默认不开启，只有用户提交了路径，才会去读取配置文件

File: lmdeploy/serve/openai/api_server.py [hunk 5]
```diff
@@ -562,12 +924,25 @@ def serve(model_path: str,
...
     VariableInterface.async_engine = AsyncEngine(model_path=model_path,
                                                  model_name=model_name,
                                                  instance_num=instance_num,
                                                  tp=tp,
                                                  **kwargs)
+    VariableInterface.qos_engine = QosEngine(instance_num=instance_num,
+                                                 qos_tag=qos_config_str,
+                                                 engine=VariableInterface.async_engine,
+                                                 **kwargs)
+    VariableInterface.qos_engine.start()
```
[the remark this answers is not in the document]
Comment (awslshadowstar) [reply lmdeploy/serve/openai/api_server.py:944]: 如果这么些会导致 访问 /v1/chat/interactive_qos 等三个请求出错的，这三个请求里有用到 VariableInterface.qos_engine 接口

File: lmdeploy/serve/qos_engine/qos_engine.py (hunk 1 above)
Comment (awslshadowstar) [reply lmdeploy/serve/qos_engine/qos_engine.py:135]: 好的好的 改成 isinstance(request,CompletionRequestQos)

File: lmdeploy/serve/openai/protocol.py [hunk 6]
```diff
@@ -141,6 +161,28 @@ class CompletionRequest(BaseModel):
...
+    stream: Optional[bool] = False
+    top_p: Optional[float] = 1.0
+    logprobs: Optional[int] = None
+    echo: Optional[bool] = False
+    presence_penalty: Optional[float] = 0.0
+    frequency_penalty: Optional[float] = 0.0
+    user: Optional[str] = None
+    # additional argument of lmdeploy
+    repetition_penalty: Optional[float] = 1.0
+    session_id: Optional[int] = -1
```
[the remark this answers is not in the document]
Comment (awslshadowstar) [reply lmdeploy/serve/openai/protocol.py:182]: 好的好的

File: lmdeploy/serve/qos_engine/qos_engine.py (hunk 2 above)
Comment (awslshadowstar) [reply lmdeploy/serve/qos_engine/qos_engine.py:32]: okk

**Comment by lvhan028** [comment]
图像不建议用文件的方式放入repo，这样会导致项目太大了。请使用github图床方式，加入图像的链接

**Comment by sallyjunjun** [comment]
> May provide user guide

**Comment by awslshadowstar** [comment]
> 图像不建议用文件的方式放入repo，这样会导致项目太大了。请使用github图床方式，加入图像的链接

现在github 图床获取到的链接都是 private-user-images.githubusercontent.com 域名的带过期jwt的图片链接，不适合用于文档，请问是否可以使用第三方图床工具

**Comment by lvhan028** [comment]
> > 图像不建议用文件的方式放入repo，这样会导致项目太大了。请使用github图床方式，加入图像的链接
> 
> 现在github 图床获取到的链接都是 private-user-images.githubusercontent.com 域名的带过期jwt的图片链接，不适合用于文档，请问是否可以使用第三方图床工具

把图像拽到这个PR下的comment栏中，会自动生成一条 url，用这个url就可以了

**Comment by lvhan028** [comment]
linting的错误还在：
```
pip install pre-commit
cd lmdeploy
pre-commit install .
pre-commit run --all-files
```
用上面这条命令，能够检查commit之前所有的文件格式规范上的错误，可以在本地修改好一并提交

**Comment by awslshadowstar** [comment]
qos_moved_to_queue_head

**Comment by lvhan028** [comment]
May improve docstring coverage of qos_engine module

File: lmdeploy/serve/qos_engine/qos_engine.py [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+
+logger = logging.getLogger(__name__)
+
+
+class QosConfig:
+
+    def __init__(self, qos_tag=''):
+        try:
+            qos_config = json.loads(qos_tag)
+            self.is_qos_enabled = qos_config['enable_user_qos']
```
Comment (lvhan028) [inline lmdeploy/serve/qos_engine/qos_engine.py:22]: 建议使用 `self.is_qos_enabled = qos_config.get('enable_user_qos', False)`
这样可以避免 try/exception

File: lmdeploy/serve/openai/api_server.py [hunk 8]
```diff
@@ -562,12 +911,25 @@ def serve(model_path: str,
...
+                qos_config_str = file.read()
+        except FileNotFoundError:
+            qos_config_str = ''
 
     VariableInterface.async_engine = AsyncEngine(model_path=model_path,
                                                  model_name=model_name,
                                                  instance_num=instance_num,
                                                  tp=tp,
                                                  **kwargs)
+    VariableInterface.qos_engine = QosEngine(
```
Comment (lvhan028) [inline lmdeploy/serve/openai/api_server.py:927]: `qos_engine` always start, why?

File: lmdeploy/serve/openai/api_server.py [hunk 9]
```diff
@@ -84,6 +86,143 @@ def ip2id(host_ip: str):
...
+        request.session_id = random.randint(1, 10086)
+    error_check_ret = await check_request(request)
+    if error_check_ret is not None:
+        return error_check_ret
+
+    model_name = request.model
+    request_id = str(request.session_id)
+    created_time = int(time.time())
+
+    result_generator = await VariableInterface.qos_engine.generate_with_qos(
```
Comment (lvhan028) [inline lmdeploy/serve/openai/api_server.py:132]: May check `qos_engine` valid or not

File: lmdeploy/serve/openai/api_server.py [hunk 10]
```diff
@@ -230,6 +369,149 @@ async def completion_stream_generator() -> AsyncGenerator[str, None]:
...
+    if error_check_ret is not None:
+        return error_check_ret
+
+    model_name = request.model
+    request_id = str(request.session_id)
+    created_time = int(time.time())
+    if isinstance(request.prompt, str):
+        request.prompt = [request.prompt]
+
+    generators = await VariableInterface.qos_engine.generate_with_qos(request)
```
Comment (lvhan028) [inline lmdeploy/serve/openai/api_server.py:420]: May check `qos_engine` valid or not

File: lmdeploy/serve/qos_engine/qos_engine.py [hunk 11]
```diff
@@ -0,0 +1,241 @@
...
+        self.engine = engine
+        self.availSlots = engine.instance_num
+        self._stop_event = threading.Event()
+        self._dequeue_thread = threading.Thread(target=self._serve,
+                                                daemon=True)
+        self.qos_config = QosConfig(qos_tag)
+
+        self.qos_user_group = QosGroupQueue(self.qos_config)
+
+        self.usage_stats = UsageStats(60, 6, 0,
```
Comment (lvhan028) [inline lmdeploy/serve/qos_engine/qos_engine.py:48]: May use `key=value` to initiate `UsageStats` in order to clarify what 60, 6, 0 are

File: lmdeploy/serve/qos_engine/qos_engine.py [hunk 12]
```diff
@@ -0,0 +1,241 @@
...
+                    self.user_served_reqs = dict()
+            logger.info(
+                f'qos svc running for {ts} seconds,last 20 seconds: {outdata}')
+            ts += 20
+            time.sleep(20)
+
+    def dequeue(self, usage_stats):
+        return self.qos_user_group.dequeue(usage_stats)
+
+    def stop(self):
```
Comment (lvhan028) [inline lmdeploy/serve/qos_engine/qos_engine.py:208]: When will `stop` be called?

File: lmdeploy/serve/qos_engine/qos_engine.py [hunk 13]
```diff
@@ -0,0 +1,241 @@
...
+        with self.lock:
+            self.qos_user_group.enqueue(request_event)
+
+        await event.wait()
+
+        result_generator = await self.generate(request)
+
+        # release self.availSlots resources
+        with self.lock:
+            if isinstance(request, CompletionRequestQos):
```
Comment (lvhan028) [inline lmdeploy/serve/qos_engine/qos_engine.py:150]: Is `request.prompt` a list or a string?

File: lmdeploy/serve/qos_engine/qos_engine.py (hunk 11 above)
Comment (awslshadowstar) [reply lmdeploy/serve/qos_engine/qos_engine.py:48]: okk

File: lmdeploy/serve/openai/api_server.py (hunk 10 above)
Comment (awslshadowstar) [reply lmdeploy/serve/openai/api_server.py:420]: if `VariableInterface.qos_engine` is not valid，`VariableInterface.qos_engine` will be a quote of  VariableInterface.async_engine

File: lmdeploy/serve/qos_engine/qos_engine.py (hunk 12 above)
Comment (awslshadowstar) [reply lmdeploy/serve/qos_engine/qos_engine.py:208]: it will be deleted

File: lmdeploy/serve/openai/api_server.py (hunk 9 above)
Comment (awslshadowstar) [reply lmdeploy/serve/openai/api_server.py:132]: okk

File: lmdeploy/serve/openai/api_server.py (hunk 8 above)
Comment (awslshadowstar) [reply lmdeploy/serve/openai/api_server.py:927]: okk

File: lmdeploy/serve/qos_engine/qos_engine.py (hunk 7 above)
Comment (awslshadowstar) [reply lmdeploy/serve/qos_engine/qos_engine.py:23]: 好的好的

File: lmdeploy/serve/qos_engine/qos_engine.py (hunk 13 above)
Comment (awslshadowstar) [reply lmdeploy/serve/qos_engine/qos_engine.py:150]: if request is  `CompletionRequestQos`, request.prompt is a list ,else a string

File: lmdeploy/serve/qos_engine/inner_group_schd.py [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+
+        total_quota = 0
+        for item in user_id_map:
+            total_quota += item['quota_pct']
+        for item in user_id_map:
+            user_id = item['id']
+            self.user_queue_map[user_id] = collections.deque()
+            self.user_quota_map[user_id] = item['quota_pct'] / total_quota
+
+        self.lock = threading.Lock()
```
Comment (lvhan028) [inline lmdeploy/serve/qos_engine/inner_group_schd.py:26]: `self.lock` is defined but not used

Comment (Dofgal) [reply lmdeploy/serve/qos_engine/inner_group_schd.py:26]: It'll be deleted

File: docs/zh_cn/qos.md [hunk 15]
```diff
@@ -0,0 +1,260 @@
...
+            {
+                "id": "default",
+                "quota_pct": "0"
+            }
+        ]
+    }
+}
+```
+
+### 如何使用 LMDeploy-QoS 感知进行推理
```
Comment (AllentDan) [inline docs/zh_cn/qos.md:130]: 这里文档好像没贴对

Comment (Dofgal) [reply docs/zh_cn/qos.md:130]: 请问具体是什么问题呢

File: docs/en/qos.md [hunk 16]
```diff
@@ -0,0 +1,260 @@
...
+
+To be noted, there could be other ways of hybrid strategies 1 &2, this doc only introduced one method that works well in our scenario. Other hybrid methods shall consider that the prioritization and pro-rated sharing are obviously conflicting strategies so there isn't an easy way to mix them to work within a single dimension.
+
+That been said, there are ways to meet users' simple requirements to only make use of either one of the strategies. For example, a system requiring only a fair-sharing-like strategy can be configured with all users under one group. A good example is one model deployed for two different business purposes without dependency. A system having need of strictly chained prioritizations between different groups of users can be configured with at most 4 groups and each group having only one user. A good example is DAG inference workload.
+
+### A Sample QoS Configuration
+
+The configuration will be specified by the `--qos_config_path` flag, and will be loaded by program upon startup.
+
+```json
```
Comment (AllentDan) [inline docs/en/qos.md:68]: This config can not be loaded

File: docs/zh_cn/qos.md (hunk 15 above)
Comment (AllentDan) [reply docs/zh_cn/qos.md:130]: 哦，没事，文档这里是先展示 lmdeploy 多出的参数。

File: docs/en/qos.md (hunk 16 above)
Comment (Dofgal) [reply docs/en/qos.md:68]: Oh we made a mistake in the json config template...

File: lmdeploy/serve/qos_engine/qos_engine.py (hunk 13 above)
Comment (AllentDan) [reply lmdeploy/serve/qos_engine/qos_engine.py:150]: CompletionRequestQos accepts `str` and `List[str]` both.

Comment (Dofgal) [reply lmdeploy/serve/qos_engine/qos_engine.py:150]: We will check if `request.prompt` is `str` or `List[str]`, instead of checking `request` type.

**Review by AllentDan — APPROVED** [review]
LGTM after left comments resolved.

File: docs/en/qos.md [hunk 17]
```diff
@@ -0,0 +1,252 @@
...
+        "Platinum",
+        "Gold",
+        "Silver",
+        "Bronze"
+    ],
+    "user_group_map": {
+        "Platinum": [
+            {
+                "id": "user_id0",
+                "quota_pct": "100"
```
Comment (AllentDan) [inline docs/en/qos.md:81]: seems `quota_pct` does not accept string

Comment (Dofgal) [reply docs/en/qos.md]: Changed to `int`

**Review by lvhan028 — APPROVED** [review]

**Pull request merged**